consensus/reactor.go

```diff
@@ -760,967 +662,720 @@ func (conR *Reactor) gossipVotesForHeight(
-	return func(conR *Reactor) { conR.Metrics = metrics }
-}
+// processPeerUpdate process a peer update message. For new or reconnected peers,
+// we create a peer state if one does not exist for the peer, which should always
+// be the case, and we spawn all the relevant goroutine to broadcast messages to
+// the peer. During peer removal, we remove the peer for our set of peers and
+// signal to all spawned goroutines to gracefully exit in a non-blocking manner.
+func (r *Reactor) processPeerUpdate(peerUpdate p2p.PeerUpdate) {
+	r.Logger.Debug("received peer update", "peer", peerUpdate.PeerID, "status", peerUpdate.Status)
 
-//-----------------------------------------------------------------------------
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
 
-var (
-	ErrPeerStateHeightRegression = errors.New("error peer state height regression")
-	ErrPeerStateInvalidStartTime = errors.New("error peer state invalid startTime")
-)
+	switch peerUpdate.Status {
+	case p2p.PeerStatusNew, p2p.PeerStatusUp:
+		// Do not allow starting new broadcasting goroutines after reactor shutdown
+		// has been initiated. This can happen after we've manually closed all
+		// peer goroutines and closed r.closeCh, but the router still sends in-flight
+		// peer updates.
+		if !r.IsRunning() {
+			return
+		}
 
-// PeerState contains the known state of a peer, including its connection and
-// threadsafe access to its PeerRoundState.
-// NOTE: THIS GETS DUMPED WITH rpc/core/consensus.go.
-// Be mindful of what you Expose.
-type PeerState struct {
-	peer   p2p.Peer
-	logger log.Logger
-
-	mtx   sync.Mutex             // NOTE: Modify below using setters, never directly.
-	PRS   cstypes.PeerRoundState `json:"round_state"` // Exposed.
-	Stats *peerStateStats        `json:"stats"`       // Exposed.
-}
+		var (
+			ps *PeerState
+			ok bool
+		)
 
-// peerStateStats holds internal statistics for a peer.
-type peerStateStats struct {
-	Votes      int `json:"votes"`
-	BlockParts int `json:"block_parts"`
-}
+		ps, ok = r.peers[peerUpdate.PeerID]
+		if !ok {
+			ps = NewPeerState(r.Logger, peerUpdate.PeerID)
+			r.peers[peerUpdate.PeerID] = ps
+		}
 
-func (pss peerStateStats) String() string {
-	return fmt.Sprintf("peerStateStats{votes: %d, blockParts: %d}",
-		pss.Votes, pss.BlockParts)
-}
+		if !ps.IsRunning() {
+			// Set the peer state's closer to signal to all spawned goroutines to exit
+			// when the peer is removed. We also set the running state to ensure we
+			// do not spawn multiple instances of the same goroutines and finally we
+			// set the waitgroup counter so we know when all goroutines have exited.
+			ps.broadcastWG.Add(3)
+			ps.SetRunning(true)
+
+			// start goroutines for this peer
+			go r.gossipDataRoutine(ps)
+			go r.gossipVotesRoutine(ps)
+			go r.queryMaj23Routine(ps)
+
+			// Send our state to the peer. If we're fast-syncing, broadcast a
+			// RoundStepMessage later upon SwitchToConsensus().
+			if !r.waitSync {
+				r.sendNewRoundStepMessage(ps.peerID)
```

Comment:
We may want to consider running this as a goroutine, since it grabs the conS.mtx lock and thus may block for a while as it's waiting for it.

Reply:
Yes, makes sense since we send on a p2p channel. Good catch!